test/JoinTest.cpp

```diff
@@ -0,0 +1,429 @@
+      JOIN_FUNCTION func,
+      ad_utility::source_location l = ad_utility::source_location::current()
+    ) {
+  // For generating better messages, when failing a test.
+  auto trace{generateLocationTrace(l, "goThroughSetOfTestsWithJoinFunction")};
+
+  for (normalJoinTest const& test: testSet) {
+    IdTable resultTable{useJoinFunctionOnIdTables(test.leftInput,
+        test.rightInput, func)};
+
+    compareIdTableWithExpectedContent(resultTable, test.expectedResult, test.resultMustBeSortedByJoinColumn, test.leftInput.joinColumn);
+  }
+}
+
+void runTestCasesForAllJoinAlgorithm(std::vector<normalJoinTest> testSet,
+    ad_utility::source_location l = ad_utility::source_location::current()) {
+  // For generating better messages, when failing a test.
+  auto trace{generateLocationTrace(l, "runTestCasesForAllJoinAlgorithm")};
+ 
+  // All normal join algorithm defined as lambda functions for easier handing
+  // over to helper functions.
+  Join J{Join::InvalidOnlyForTestingJoinTag{}};
+  auto HashJoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.hashJoin<A, B, C>(AD_FWD(args)...);
+  };
+  auto JoinLambda = [&J]<int A, int B, int C>(auto&&... args) {
+    return J.join<A, B, C>(AD_FWD(args)...);
+  };
+
+  // For sorting IdTables by their join column.
+  auto sortByJoinColumn = [](IdTable& idTable, size_t joinColumn) {
+    /*
+     * TODO Introduce functionality to the IdTable-class, so that std::ranges::sort
+     * can be used instead of std::sort. Currently it seems like the iterators
+     * , produced by IdTable, aren't the right type.
+     */
+    std::sort(idTable.begin(), idTable.end(), [&joinColumn](const auto& row1,
+          const auto& row2) {return row1[joinColumn] < row2[joinColumn];});
+  };
+
+  // Random shuffle both tables, run hashJoin, check result.
+  for (normalJoinTest& testCase: testSet) {
+    randomShuffle(testCase.leftInput.idTable.begin(), testCase.leftInput.idTable.end());
+    randomShuffle(testCase.rightInput.idTable.begin(), testCase.rightInput.idTable.end());
+    testCase.resultMustBeSortedByJoinColumn = false;
+  }
+  goThroughSetOfTestsWithJoinFunction(testSet, HashJoinLambda);
+
+  // Sort the larger table by join column, run hashJoin, check result (this time it's sorted).
+  for (normalJoinTest& testCase: testSet) {
+    if (testCase.leftInput.idTable.size() >=
+        testCase.rightInput.idTable.size()) {
+      sortByJoinColumn(testCase.leftInput.idTable, testCase.leftInput.joinColumn);
+    } else {
+      sortByJoinColumn(testCase.rightInput.idTable, testCase.rightInput.joinColumn);
+    }
+
+    testCase.resultMustBeSortedByJoinColumn = true;
+  }
+  goThroughSetOfTestsWithJoinFunction(testSet, HashJoinLambda);
+
+  // Sort both tables, run merge join and hash join, check result. (Which has to be sorted.)
+  for (normalJoinTest& testCase: testSet) {
+    sortByJoinColumn(testCase.leftInput.idTable, testCase.leftInput.joinColumn);
```

Comment:
You can add an overload (or change the function, if the other one is not needed), 
s.t. you can directly write `sortByJoinColumn(testCase.leftInput)` (can also be used above).

Reply:
Already had that idea and implemented it.

Looking back, it was kinda stupid, that I didn't write the first version like that. I mean, the inputs are a special type that holds the IdTable and the number of it's join column, so why not use that?